### DISCUSSION TITLE
how to program math formula in 'extra devices'(need help!)

### DISCUSSION THREAD

**Discussion by phillip-kang** [artisan-roaster-scope/artisan#1526] [Q&A]
Hello,
I'm working on a research regarding the derivative of delta ET(in short, DDET).

Here's what I have so far;
The pink line shows the value of DDET at each point, where y=R1-R1[-1].
Black straight line shows where "delta ET = 0".

![a1](https://github.com/artisan-roaster-scope/artisan/assets/79606762/cd370339-1197-477b-a8df-fa157e83e371)

The next step I'm going to do is to calculate each area of those minus DDET values. (circled in red)
I want to know how big those areas are and show it on the graph.

But creating formula is way more trickier than I thought:(

Preconditions of the formula are as follows;
1) when calculating cumulative DDET values, only negative values are to be considered. positive values must be ruled out.
2) when DDET is positive, the output of the formula is "0"

For those who have good skills in computer programming, please help me:)
Thanks!

**Comment by MAKOMO** [comment]
> Hello, I'm working on a research regarding the derivative of delta ET(in short, DDET).
> 
> Here's what I have so far; The pink line shows the value of DDET at each point, where y=R1-R1[-1]. 

That is a simplified version of a derivative taking assuming the time deltas to be constant.
> 
> Preconditions of the formula are as follows;
> 
> 1. when calculating cumulative DDET values, only negative values are to be considered. positive values must be ruled out.
> 2. when DDET is positive, the output of the formula is "0"
> 
> For those who have good skills in computer programming, please help me:) Thanks!

My suggestion to the above would be the following

```
(0 if (t == 0) else F1+abs(min(0,R1 - R1[-1])))
````

The `min(0,x)` evaluates to 0 for all positive `x`. For the negative values we switch the sign by taking the absolute value `abs(x)`. Then we add the result to the previous value hold in  `F1` to accumulate those (now positive) values. Note that the values that were originally positive got substituted by that `0` and adding a `0` to anything does not change its value. Thus those originally positive DDET readings do not contribute to our cumulative sum. Now for this commutative sum to work we need to initalize it. Here we initialize our sum to 0 by the if-then-else statement which defines the start at  time=0.

Hope that the resulting curve makes any sense to you....

**Reply by phillip-kang** [reply]
Thank you so much!

But the function you gave me is presenting upward slope.

(0 if (t == 0) else F1+abs(min(0,R1 - R1[-1])))

And here's the result.
I copy & pasted the line above and updated it.
(Blue line)

![a](https://github.com/artisan-roaster-scope/artisan/assets/79606762/607c220f-3d52-4621-896a-a0b047addd39)

What I first meant was the graph like;
 1) When DDET<0, convert all the negative values to positive, make the cumulative sum and show it on the graph.
    => This worked so well

 2) When DDET >=0, initiate to "0".
    => Your function is taking an upward slope.
    => I would like the previous value to be initialized when it meets the condition 'DDET>=0'.
    => **PLUS, if I can elaborate this condition, I would also like to ignore a sudden noise of DDET.**
          
        For example,
        While DDET graph is in negative phase, 
        and the graph suddenly exhibits zero or positive values during a short period of time(like t1 or t1 ~ t2),
        It's likely to be the reading noise and I'd like to ignore in that it should not activate the initiating condition.
        In this case, the previous sum still maintains the value and adds up to the following absolute values of negatives.
         
 3) and repeat those two steps.

Then my suggestion will probably look like this.
![a rev](https://github.com/artisan-roaster-scope/artisan/assets/79606762/0da90162-850e-47f0-a5ef-90baa64d9008)

If you see thick blue line I drew(sorry for clumsy drawing), 
the result on Y value goes up only when DDET<0 (previous absolute values of negative DDETs continuously being added up),
then drops to "0" when DDET>=0 (value initialized)
and repeat those processes. 

Hope my explanation helped your understanding.

Thanks!

**Comment by MAKOMO** [comment]
> 1. calculating cumulative DDET values

> Thank you so much!
> 
> But the function you gave me is presenting upward slope.

Yes, because it is cumulative. That is what you wanted, right: _...calculating cumulative DDET values..._
> 
> (0 if (t == 0) else F1+abs(min(0,R1 - R1[-1])))
> 
> And here's the result. I copy & pasted the line above and updated it. (Blue line)
> 

as expected
> 
> What I first meant was the graph like;

but not what you wrote

> 
> 1. When DDET<0, convert all the negative values to positive, make the cumulative sum and show it on the graph.
>    => This worked so well
> 2. When DDET >=0, initiate to "0".
>    => Your function is taking an upward slope.
>    => I would like the previous value to be initialized when it meets the condition 'DDET>=0'.
>    => **PLUS, if I can elaborate this condition, I would also like to ignore a sudden noise of DDET.**
>    ```
>    For example,
>    While DDET graph is in negative phase, 
>    and the graph suddenly exhibits zero or positive values during a short period of time(like t1 or t1 ~ t2),
>    It's likely to be the reading noise and I'd like to ignore in that it should not activate the initiating condition.
>    In this case, the previous sum still maintains the value and adds up to the following absolute values of negatives.
>    ```
> 3. and repeat those two steps.
> 

Just drop the cumulative summing and use

```
abs(min(0,R1 - R1[-1]))
```

**Reply by phillip-kang** [reply]
Sorry to bother you.
Since I'm not a native speaker, I'm having trouble understanding your words, as well as expressing my intention.

At the very end of your comment, you said; just use "abs(min(0,R1 - R1[-1]))".
It is the converted numbers of DDET, not the partial integration. Right?

To repeat, here is the math function I want to design; 
1) I like to see how big negative areas are **respectively**,  I didn't mean the total sum of each negative areas.
    -> According to your first suggestion, the graph keeps going up and does not project how big each negative area is.
    -> The summing algorithm must work only when DDET<0.
         if DDET>=0, the graph should go back to zero so that we can start to calculate sum in the next negative session.
    -> This way I can separately see each area on the graph and control heat.
2) While we are defining the negative areas in number 1 above, 
     few of the reading noise(sudden positive sign or zero during a single or couple of time units) in the negative phase
     can work as interference, since It will make the previous sum value suddenly "zero".
    
Then the graph will probably look like this(thick blue line),
the graph(cumulation of negative values) goes up when DDET<0, then zero when DDET>=0, up and down and so forth

This make it much easier to compare the areas of each negative phase,
(which means the total amount of evaporation for me)
and make heat control.

![image](https://github.com/artisan-roaster-scope/artisan/assets/79606762/eb6c9b73-2876-47ed-a0ca-132c52f04303)

If you did not understand my explanation or have any question,
please let me know.

Thanks.

**Reply by MAKOMO** [reply]
My previous response had the formula without the cumulative sum in the last line

**Reply by phillip-kang** [reply]
Dear Marko,
Do you mind if I have the entire formula once again with all above conditions combined?

Thanks

**Reply by poundy** [reply]
> Dear Marko, Do you mind if I have the entire formula once again with all above conditions combined?
> 
> Thanks

how hard is scrolling up on the github response??

**Reply by phillip-kang** [reply]
Man,
is that what you say to someone who ask for help?
I said I still didn't understand his entire context, not that I got tedious for moving my mouse:(
Do you really mean that I wanted to bother him without even using the translator?

Dear Marko,
I'm sorry, but here's what I understood from scratch.

1) Since I inquired the cumulative formula, your reply was this;

- (0 if (t == 0) else F1+abs(min(0,R1 - R1[-1]))))

2) After adapting and updating, the graph was showing an upward slope, it keeps being cumulative.
    (blue line)
![image](https://github.com/artisan-roaster-scope/artisan/assets/79606762/9c1c6ed5-cf88-481e-b322-ef313859d188)

   The graph I wanted in the first place was to show the areas of each negative signs so that I can compare the scale each other.
   In addition to this, I wanted  to add noise process that shows positive or zero for a short period of negative phase. 
   Sketching these ideas, It would probably look like this.
   (Thick blue line)

   You see the converted numbers being added up until the next positive sign begins, and then turn back to zero when positive shows 
   up, and build up numbers again in negative.(while positive noise being ignored) 
![image](https://github.com/artisan-roaster-scope/artisan/assets/79606762/4a4d0411-83eb-4694-be49-fd9d4f40a851)

3) And then, your final reply to number 2 was to use "abs(min(0,R1 - R1[-1]))" while deleting F1(cumulative sum),
    which makes;

- 0 if (t == 0) else abs(min(0,R1 - R1[-1]))

 and the graph only shows converted signs of negative values, while making others signals zero.
 But it didn't reflect the requirements of summing, noise process, as well as nullification.
![image](https://github.com/artisan-roaster-scope/artisan/assets/79606762/e61f44a5-097a-4420-a62d-d5cf1f85319d)

This is why I got lost in this conversation.
Thank you so much.

**Reply by MAKOMO** [reply]
All your requirements are unclear to me und YOURS. You also never clearly indicated where your requirements are coming from and what the resulting signal should be good for. I am out. My advice is to do whatever you want to do outside of Artisan eg. in Excel or your own program. Artisan is a roast logger and not a general signal processor.

Sorry, this goes beyond on what I want to spend my time on. Maybe someone else?

**Reply by poundy** [reply]
Yes, that's what I would tell someone who clearly missed a point but never looked twice...
Artisan is a community project, and is supported from community donations and donations from Manufacturers / Suppliers who wish to support the project. You clearly have some idea what you are doing all this for, and you haven't shared anything with us that would help us try to understand what or why you're trying to do it, and as Marko says it's beyond the sensible scope for a roast logging system.  There's lots of places you could do these calculations before it hits Artisan if you think it's worthwhile.

**Reply by phillip-kang** [reply]
Explaining the background is not a big deal, but for me it takes about a couple of diagrams and conversation to go through,
which won't fit to this kind of long thread and lots of work.

I understood.
Forgive me I was a little carried away.

Dear Marko,
I'd be glad if you can recommend me any website other than Artisan webpage,
in which you shared on your own detail information on math language used in Artisan. e.g. home-barista.com

In case I might as well manage this matter myself.
Thanks for your support.

**Comment by roasterdave** [comment]
>won't fit this kind of long thread and lots of work.

Lots of time and work to were spent here trying to answer your question.  It would be kind of you to take the time to explain your project.

>I'd be glad if you can recommend me any website other than Artisan webpage,

The very best resource on symbolic equations is found in the Artisan blog.
[Symbolic Formulas: Basics, New Variables and Applications](https://artisan-roasterscope.blogspot.com/2019/11/symbolic-formulas-basics-new-variables.html)
[Signals, Symbolic Assignments and the Plotter](https://artisan-roasterscope.blogspot.com/2015/10/signals-symbolic-assignments-and-plotter.html)
[Roast Calculus](https://artisan-roasterscope.blogspot.com/2016/03/roast-calculus.html)
[Search the blog for "symbolic"](https://artisan-roasterscope.blogspot.com/search?q=symbolic)

**Reply by phillip-kang** [reply]
Thanks Dave,
I appreciate your kind reply.
